p2panda-rs/src/atomic/entry.rs

```diff
@@ -0,0 +1,168 @@
+    pub fn sign_and_encode(&self, key_pair: &KeyPair) -> Result<EntryEncoded> {
+        // Generate message hash
+        // @TODO: Handle case where message is not set
+        let message_encoded = self.message.clone().unwrap().encode().unwrap();
+        let message_hash = message_encoded.hash();
+        let message_size = message_encoded.size();
+
+        // Convert entry links to Bamboo crate types
+        let backlink = self
+            .entry_hash_backlink
+            .clone()
+            .map(|link| YamfHash::Blake2b(link.to_bytes()));
+
+        let lipmaa_link = self
+            .entry_hash_skiplink
+            .clone()
+            .map(|link| YamfHash::Blake2b(link.to_bytes()));
+
+        // Create bamboo entry. See: https://github.com/AljoschaMeyer/bamboo#encoding for encoding
+        // details and definition of entry fields.
+        let mut entry: BambooEntry<_, &[u8]> = BambooEntry {
+            log_id: self.log_id.as_u64(),
+            is_end_of_feed: false,
+            payload_hash: YamfHash::Blake2b(message_hash.to_bytes()),
+            payload_size: message_size,
+            author: PublicKey::from_bytes(&key_pair.public_key_bytes())?,
+            seq_num: self.seq_num.as_u64(),
+            backlink,
+            lipmaa_link,
+            sig: None,
+        };
+
+        // @TODO: Sign BambooEntry and encode it
+        EntryEncoded::new("dummy")
+    }
+
+    /// Decodes an encoded entry and returns it.
+    pub fn from_encoded(entry_encoded: EntryEncoded) -> Self {
+        entry_encoded.decode()
+    }
+
+    /// Returns hash of backlink entry when given.
+    pub fn backlink_hash(&self) -> Option<Hash> {
+        self.entry_hash_backlink.clone()
+    }
+
+    /// Returns hash of skiplink entry when given.
+    pub fn skiplink_hash(&self) -> Option<Hash> {
+        self.entry_hash_skiplink.clone()
+    }
+
+    /// Returns sequence number of entry.
+    pub fn seq_num(&self) -> SeqNum {
+        self.seq_num.clone()
+    }
+
+    /// Calculates sequence number of backlink entry.
+    pub fn seq_num_backlink(&self) -> Option<SeqNum> {
+        self.seq_num.backlink_seq_num()
+    }
+
+    /// Calculates sequence number of skiplink entry.
+    pub fn seq_num_skiplink(&self) -> Option<SeqNum> {
+        self.seq_num.skiplink_seq_num()
+    }
+
+    /// Returns true if entry contains message.
+    pub fn has_message(&self) -> bool {
+        self.message.is_some()
+    }
+}
+
+impl Validation for Entry {
+    fn validate(&self) -> Result<()> {
+        // First entries do not contain any sequence number or links. Every other entry has to
+        // contain all information.
+        if (self.entry_hash_backlink.is_none()
+            && self.entry_hash_skiplink.is_none()
+            && self.seq_num.is_first())
+            || (self.entry_hash_backlink.is_some()
+                && self.entry_hash_skiplink.is_some()
+                && !self.seq_num.is_first())
```

Comment:
Isn't this the wrong way around? Validation should throw an error for the sequence number = 1 case when backlink and skiplink **are** defined, not the other way?

Reply:
Ha!

Reply:
Well spotted! I guess some tests are missing 🤭